### PR TITLE
docs(research): #1831 M3 math spec — qwen35moe HIP decode semantics from llama.cpp

### PR DESCRIPTION
**Pull request by bong-water-water-bong**
### **User description**
**Part of #1831** (open). M3 kickoff artifact: `docs/research/hip-qwen35-gdn-m3-spec.md` pins the exact decode math for the HIP kernels, sourced from the authoritative llama.cpp implementation (`ggml-org/llama.cpp/src/models/qwen35moe.cpp`, the writer/loader of the `qwen35moe` GGUF family) + the Q8_0 header capture already on main.

**Resolves the M2 port-map open questions:**
- Full-attn layers (10, (il+1)%4==0): 8192-row `attn_q` = per-head [q 256 | gate 256] pairs; per-head q/k RMSNorm; **IMRoPE** multi-section rope [11,11,10,0] (≠ the dense backend's simple rope — new kernel); sigmoid-gated attention; `attn_output` 4096→H.
- GDN layers (30): `attn_qkv` rows = [q 2048 | k 2048 | v 4096] (head_k_dim 128×16 k-heads, head_v_dim 128×32 v-heads); `attn_gate` = z (4096); decay gate = ssm_a·softplus(ssm_alpha@x + dt_bias); beta = sigmoid(ssm_beta@x); conv1d(4) over 8192 ch + silu; per-head L2 q/k; k-repeat 16→32; delta rule over state (128×128×32/seq); z-gate group-RMSNorm × silu(z); `ssm_out` 2048×4096.
- MoE (every layer): softmax router top-8 over 256; fused 3-D expert tensors with per-expert 512-wide slices; shared expert × sigmoid(`ffn_gate_inp_shexp`@x).

**Method note:** items marked ⚠ (exact fused-tensor channel interleave, IMRoPE bit layout, k-repeat mapping, expert slice indexing) are locked by an **empirical llama.cpp debug-tensor capture on strixhalo** before any kernel is written — the repo's standing no-guessing rule. Kernel inventory + validation gate (corr ≥ 0.99 vs CPU over 100+ tokens) included.

Docs-only. No behavior change. Next: M3 capture, then GDN kernels.

___

### **PR Type**
Documentation

___

### **Description**
- Add M3 math spec for qwen35moe HIP decode

- Documents full-attention, GDN, and MoE math

- Records strixhalo M3CAP tensor-layout confirmations

- Defines kernel inventory and correlation validation gate

___

### Diagram Walkthrough

```mermaid
flowchart LR
  llama["llama.cpp qwen35moe.cpp"] --> spec["docs/research/hip-qwen35-gdn-m3-spec.md"]
  capture["strixhalo M3CAP tensor capture"] --> spec
  spec --> kernels["HIP kernels for #1831"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>hip-qwen35-gdn-m3-spec.md</strong><dd><code>Add M3 math spec for qwen35moe HIP decode</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/research/hip-qwen35-gdn-m3-spec.md

<ul><li>Adds 201-line M3 math spec for qwen35moe HIP decode<br> <li> Pins hyperparams and layer dispatch from llama.cpp<br> <li> Details full-attention, GDN recurrence, and MoE math<br> <li> Records M3CAP tensor-layout findings and kernel plan</ul>

</details>

  </td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2126/files#diff-c7ba3ce6b0fe0e1240990d472f77e60c2449a6b3d8d5e634bb88ce5bb5e39de0">+201/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by bong-water-water-bong** [comment]
Updated: M3 capture findings (llama.cpp qwen35moe CPU decode broken upstream — heap OOB on master + v0.4.0, gdb-verified; the file itself is sound). Source-locked layout record replaces the ⚠ items where provable; the fused-row channel-order crux + the corr-gate golden move to the in-repo CPU reference (tools/qwen35moe_cpu_ref.cpp), validated by math-family parity on the 80B q4nx file.
